packages/client/hmi-client/src/utils/petri-net-validator.ts

```diff
@@ -22,35 +22,30 @@ interface Output {
 	os: number; // State ID which is the target
 }
 
-/* Validates petrinet - check #2 must come before check #5 to avoid an infinite loop */
-export const petrinetValidator = (petrinet: Petrinet): boolean => {
+/* Validates petrinet - check #2 must come before check #3 to avoid an infinite loop */
+export const petriNetValidator = (petrinet: PetriNet): string => {
 	const { S, T, I, O } = petrinet;
 	// console.log(petrinet);
 
 	/* ----- 1. Requires at least one edge ----- */
-	if (I.length < 1 && O.length < 1) {
-		console.log('#1');
-		return false;
-	}
+	if (I.length < 1 && O.length < 1) return 'Invalid petri net: #1. Requires at least one edge';
 
 	/* ----- 2. Check that every node is at least either a source or a target ----- */
 	const checkIfSourceOrTarget = (linkedIDs: number[], lastNodeID: number): boolean => {
-		for (let id = 1; id < lastNodeID; id++) {
-			if (!linkedIDs.includes(id)) {
-				console.log('#2');
-				return false;
-			}
-		}
+		for (let id = 1; id < lastNodeID; id++) if (!linkedIDs.includes(id)) return false;
 		return true;
 	};
 	const linkedTransitionIDs: number[] = [
 		...new Set([...I.map((input) => input.it), ...O.map((output) => output.ot)])
 	];
-	if (!checkIfSourceOrTarget(linkedTransitionIDs, T.length + 1)) return false;
 	const linkedStateIDs: number[] = [
 		...new Set([...I.map((input) => input.is), ...O.map((output) => output.os)])
 	];
-	if (!checkIfSourceOrTarget(linkedStateIDs, S.length + 1)) return false;
+	if (
+		!checkIfSourceOrTarget(linkedTransitionIDs, T.length + 1) ||
+		!checkIfSourceOrTarget(linkedStateIDs, S.length + 1)
+	)
+		return 'Invalid petri net: #2. Every transition node should be at least either a source or a target';
```

Comment:
I think it'll be a good idea to remove these number codes from the return strings, `#2`, `#3`, etc.

If they're to be displayed to the user, the number will be meaningless and perhaps confusing.
If they're for the programmer, they can just Cmd+F the error message.
And any automatic conditionals we do that rely on this validator don't care about the actual content of the string.